### ISSUE TITLE
BUG: Can't install latest 1.10.0 from deb fiile

### ISSUE THREAD

**Issue by furai** [beekeeper-studio/beekeeper-studio#593]
**Describe the bug**
Can't install latest 1.10.0 from deb file

**To Reproduce**
Steps to reproduce the behaviour:
1. Have `/opt` on different drive than `/usr`
2. Try to install beekeeper through apt.
3. Watch it fail due to hardlink error between different drives.

**Expected behaviour**
Beekeeper should install properly.

**Log**
```
Preparing to unpack .../beekeeper-studio_1.10.0_amd64.deb ...
Unpacking beekeeper-studio (1.10.0) over (1.9.4) ...
13dpkg: error processing archive /var/cache/apt/archives/beekeeper-studio_1.10.0
_amd64.deb (--unpack):
 error creating hard link './usr/share/icons/hicolor/24x24/apps/beekeeper-studio
.png': Invalid cross-device link
Errors were encountered while processing:
 /var/cache/apt/archives/beekeeper-studio_1.10.0_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Version Information (please complete the following information):**
 - OS: Ubuntu 20.04 64-bit (5.10.24 kernel)
 - App Version: 1.10.0
 
 **Possible solution**
Make deb use symlinks instead of hardlinks.

**Comment by rathboma** [comment]
Thank you SO much for reporting this. I was about to do a wider rollout and you may have saved me a big firedrill!

I think this is because of the upgrade to Electron 9 (electron-builder's wrapper process does the symlinking itself).

**Comment by rathboma** [comment]
Hmm, so I took a look. I don't think this behavior has changed at all. Did you change some device configurations recently?

Can you try the `.AppImage` instead and see if that works ok?

**Comment by rathboma** [comment]
Looks like the same problem as [this issue with Zap Desktop](https://github.com/LN-Zap/zap-desktop/issues/2089).

My guess if you have a different drive mounted for /usr and /var?

**Comment by furai** [comment]
> My guess if you have a different drive mounted for /usr and /var?

Yes, that's what I meant when I was writing this post. Sorry for not being clear about it and using wrong technical terms.

**Comment by jdpiguet** [comment]
Problem is still present with 1.10.1:
`E: /var/cache/apt/archives/beekeeper-studio_1.10.1_amd64.deb: error creating hard link './usr/share/icons/hicolor/24x24/apps/beekeeper-studio.png': Invalid cross-device link
`

**Comment by rathboma** [comment]
Unfortunately my builds are done by a third party tool - Electron Builder, and I don't have a lot of control of the DEB post-install scripts.

My best immediate suggestion would be to download and use the AppImage version -- this version still allows updating with a notification.

I'll raise an issue on electron-builder

**Comment by rathboma** [comment]
Issue created with electron builder - https://github.com/electron-userland/electron-builder/issues/5721

**Comment by Rafaelmignoni** [comment]
I have the same problem. Any solutions for this?

**Comment by jdpiguet** [comment]
Just got 1.11.2 in the repositories.
Problem still present:

`root@amroth:~# apt upgrade 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
The following packages will be upgraded:
  beekeeper-studio
1 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/70.9 MB of archives.
After this operation, 13.1 MB of additional disk space will be used.
Do you want to continue? [Y/n] 
(Reading database ... 377551 files and directories currently installed.)
Preparing to unpack .../beekeeper-studio_1.11.2_amd64.deb ...
Unpacking beekeeper-studio (1.11.2) over (1.9.6) ...
dpkg: error processing archive /var/cache/apt/archives/beekeeper-studio_1.11.2_amd64.deb (--unpack):
 error creating hard link './usr/share/icons/hicolor/24x24/apps/beekeeper-studio.png': Invalid cross-device link
Errors were encountered while processing:
 /var/cache/apt/archives/beekeeper-studio_1.11.2_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)
root@amroth:~# 
`

**Comment by jdpiguet** [comment]
As workaround, one can rebuild the package with the following:

`mkdir /tmp/beekeper
dpkg-deb -R beekeeper-studio_1.11.3_amd64.deb /tmp/beekeper
cd /tmp/beekeper
for f in $(find usr -name beekeeper-studio.png); do
  s=${f##*hicolor/}
  s=${s%%/*}
  ln='opt/Beekeeper Studio/resources/public/icons/png/'${s}.png
  echo -e "\n  ${f}\n  ${ln}"
  pushd ${f%/*}
  rm beekeeper-studio.png
  ln -s "../../../../../../${ln}" beekeeper-studio.png
  popd
done
cd
dpkg-deb -b /tmp/beekeper beekeeper-studio_1.11.3~jdpiguet1_amd64.deb

dpkg -i beekeeper-studio_1.11.3~jdpiguet1_amd64.deb
`

**Comment by jdpiguet** [comment]
As a workaround, one can use the lines below:

    mkdir /tmp/beekeper
    dpkg-deb -R beekeeper-studio_1.11.3_amd64.deb /tmp/beekeper
    cd /tmp/beekeper
    for f in $(find usr -name beekeeper-studio.png); do
      s=${f##*hicolor/}
      s=${s%%/*}
      ln='opt/Beekeeper Studio/resources/public/icons/png/'${s}.png
      echo -e "\n  ${f}\n  ${ln}"
      pushd ${f%/*}
      rm beekeeper-studio.png
      ln -s "../../../../../../${ln}" beekeeper-studio.png
      popd
    done
    cd
    dpkg-deb -b /tmp/beekeper beekeeper-studio_1.11.3~jdpiguet1_amd64.deb
    
    dpkg -i beekeeper-studio_1.11.3~jdpiguet1_amd64.deb

**Comment by furai** [comment]
This is still present in the latest installer. Anyone tried fixing it?

**Comment by rathboma** [comment]
It's something to do with the post install script provided by electron-builder. I haven't had time to look at it, but in theory providing a custom version and changing how it links icons would fix it.

**Comment by zergmk2** [comment]
I met the same issue and found this post in google...

**Comment by hfelici** [comment]
You can also as root, go to /usr/share/icons/hicolor/24x24 , and there move the apps file to /opt and link it

```
cd /usr/share/icons/hicolor/24x24
mv apps /opt
ln -s /opt/apps
```

then install the dpkg and then undo the link

```
cd /usr/share/icons/hicolor/24x24
rm apps
mv /opt/apps .
```

**Comment by futu2** [comment]
Encounter the same issue.

**Comment by zeigerpuppy** [comment]
The workaround from @jdpiguet works well, thanks!
- https://github.com/beekeeper-studio/beekeeper-studio/issues/593#issuecomment-842586645

But definitely an upstream issue, electron-builder should not be using hardlinks!

**Comment by tom-ch1** [comment]
why hardlink? If you just softlink, there shouldn't be any issues.

If you can't fix such a simple issue, this is not my product, sorry
